gemato/openpgp.py

```diff
@@ -588,23 +588,46 @@ def refresh_keys_wkd(self):
             data,
             raise_on_error=OpenPGPKeyRefreshError)
 
-        # we need to explicitly ensure all keys were fetched
+        # Need to explicitly ensure all keys were fetched
+        # However:
+        # - any key MAY appear 0 or more times.
+        # - expected keys SHOULD be present.
+        # - unexpected keys MAY also be present.
+        #
+        # Track updated & unexpected keys to compare vs expected keys.
+        updated_keys = set()
+        unexpected_keys = set()
         for line in out.splitlines():
             if line.startswith(b'[GNUPG:] IMPORT_OK'):
                 fpr = line.split(b' ')[3].decode('ASCII')
                 logging.debug(
-                    f'refresh_keys_wkd(): import successful for key: {fpr}')
+                    f'refresh_keys_wkd(): import successful for key: '
+                    f'{fpr}')
                 if fpr in keys:
-                    keys.remove(fpr)
+                    updated_keys.add(fpr)
                 else:
-                    # we need to delete unexpected keys
-                    exitst, out, err = self._spawn_gpg(
-                        [GNUPG, '--batch', '--delete-keys', fpr],
-                        raise_on_error=OpenPGPKeyRefreshError)
-        if keys:
+                    # Did not expect this key, so track it to remove after
+                    # processing the import output.
+                    unexpected_keys.add(fpr)
+
+        # 128x 40-byte fingerprints = 5KiB commandline max
+        # If this contains a lot of keys, it should just blow up, but that
+        # saves complexity.
+        if unexpected_keys:
+            # we need to delete unexpected keys
+            logging.info(
+                f'refresh_keys_wkd(): got unexpected key, will remove: '
+                f'{unexpected_keys}')
+            exitst, out, err = self._spawn_gpg(
+                [GNUPG, '--batch', '--delete-keys'] + list(unexpected_keys),
+                raise_on_error=OpenPGPKeyRefreshError)
+
+        # Check if any of the keys were NOT updated
+        not_updated_keys = set(keys).difference(updated_keys)
```

Comment:
`keys` should be a set already at this point.

Reply:
cleaned up.